File: src/main/java/ru/citeck/ecos/uiserv/domain/menu/repo/MenuRepository.java
package ru.citeck.ecos.uiserv.domain.menu.repo;

import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.stereotype.Repository;

import java.time.Instant;
import java.util.List;
import java.util.Optional;
import java.util.Set;

@Repository
public interface MenuRepository extends JpaRepository<MenuEntity, Long> {
    Optional<MenuEntity> findByExtId(String extId);

    @Query("SELECT menu " +
        "FROM MenuEntity menu " +
        "JOIN menu.authorities authority " +
        "WHERE authority = ?1 " +
        "ORDER BY menu.priority DESC")
    List<MenuEntity> findAllByAuthoritiesContains(String authority);

    void deleteByExtId(String extId);

    @Query("SELECT max(m.lastModifiedDate) FROM MenuEntity m")
    Optional<Instant> getLastModifiedTime();

    @Query(
        value = "SELECT DISTINCT authority FROM ecos_menu_authority",
        nativeQuery = true
    )
    Set<String> getAllAuthoritiesWithMenu();
}

File: src/main/java/ru/citeck/ecos/uiserv/domain/menu/repo/MenuRepository.kt
package ru.citeck.ecos.uiserv.domain.menu.repo

import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.JpaSpecificationExecutor
import org.springframework.data.jpa.repository.Query
import org.springframework.stereotype.Repository
import java.time.Instant

@Repository
interface MenuRepository : JpaRepository<MenuEntity, Long>, JpaSpecificationExecutor<MenuEntity> {

    fun findByExtId(extId: String): MenuEntity?

    @Query(
        "SELECT menu " +
            "FROM MenuEntity menu " +
            "JOIN menu.authorities authority " +
            "WHERE lower(authority) = ?1 AND (menu.workspace IS NULL OR menu.workspace = '') " +
            "ORDER BY menu.priority DESC"
    )
    fun findAllByAuthoritiesContains(authority: String): List<MenuEntity>

    @Query(
        "SELECT menu " +
            "FROM MenuEntity menu " +
            "JOIN menu.authorities authority " +
            "WHERE lower(authority) = ?1 AND menu.workspace = ?2 " +
            "ORDER BY menu.priority DESC"
    )
    fun findAllByAuthoritiesContains(authority: String, workspace: String): List<MenuEntity>

    fun deleteByExtId(extId: String)

    @Query("SELECT max(m.lastModifiedDate) FROM MenuEntity m")
    fun getLastModifiedTime(): Instant?

    @Query(value = "SELECT DISTINCT authority FROM ecos_menu_authority", nativeQuery = true)
    fun getAllAuthoritiesWithMenu(): Set<String>
}
